File: sdk/node/src/highwater.js
import { VerificationError } from "./errors.js";
import { VerificationCode } from "./offline.js";

export class HighWaterGuard {
  #store;
  #allowedRollbackMs;
  #queue = Promise.resolve();

  constructor({ store, allowedRollbackMs = 0 }) {
    if (!store || typeof store.load !== "function" || typeof store.save !== "function") {
      throw new TypeError("ynlicense: high-water store with load and save is required");
    }
    if (!Number.isFinite(allowedRollbackMs) || allowedRollbackMs < 0) {
      throw new TypeError("ynlicense: allowedRollbackMs cannot be negative");
    }
    this.#store = store;
    this.#allowedRollbackMs = allowedRollbackMs;
  }

  checkAndUpdate(current = new Date()) {
    const operation = this.#queue.then(() => this.#checkAndUpdate(current));
    this.#queue = operation.catch(() => undefined);
    return operation;
  }

  async #checkAndUpdate(current) {
    const now = parseDate(current, "current time");
    const loaded = await this.#store.load();
    const last = loaded == null || loaded === "" ? null : parseDate(loaded, "saved high-water time");
    if (last) {
      if (now.getTime() + this.#allowedRollbackMs < last.getTime()) {
        throw new VerificationError(VerificationCode.CLOCK_ROLLBACK, "current time is earlier than the saved high-water time");
      }
      if (now.getTime() <= last.getTime()) return;
    }
    await this.#store.save(now);
  }
}

function parseDate(value, name) {
  const date = value instanceof Date ? new Date(value.getTime()) : new Date(value);
  if (Number.isNaN(date.getTime())) throw new TypeError(`ynlicense: ${name} is invalid`);
  return date;
}

File: sdk/go/highwater.go
package ynlicense

import (
	"context"
	"errors"
	"sync"
	"time"
)

type HighWaterStore interface {
	Load(ctx context.Context) (time.Time, error)
	Save(ctx context.Context, value time.Time) error
}

type HighWaterGuard struct {
	store           HighWaterStore
	allowedRollback time.Duration
	mu              sync.Mutex
}

func NewHighWaterGuard(store HighWaterStore, allowedRollback time.Duration) (*HighWaterGuard, error) {
	if store == nil {
		return nil, errors.New("ynlicense: high-water store is required")
	}
	if allowedRollback < 0 {
		return nil, errors.New("ynlicense: allowed rollback cannot be negative")
	}
	return &HighWaterGuard{store: store, allowedRollback: allowedRollback}, nil
}

func (g *HighWaterGuard) CheckAndUpdate(ctx context.Context, current time.Time) error {
	if current.IsZero() {
		return errors.New("ynlicense: current time is required")
	}
	g.mu.Lock()
	defer g.mu.Unlock()
	current = current.UTC()
	last, err := g.store.Load(ctx)
	if err != nil {
		return err
	}
	if !last.IsZero() {
		last = last.UTC()
		if current.Add(g.allowedRollback).Before(last) {
			return verificationError(VerificationClockRollback, "current time is earlier than the saved high-water time")
		}
		if !current.After(last) {
			return nil
		}
	}
	return g.store.Save(ctx, current)
}
